src/analysis/normed_space/box_subadditive.lean

```diff
@@ -0,0 +1,541 @@
+/-
+Copyright (c) 2020 Yury G. Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Yury G. Kudryashov
+-/
+import data.real.ennreal
+import topology.metric_space.basic
+import linear_algebra.affine_space.ordered
+import analysis.normed_space.add_torsor
+import analysis.specific_limits
+import analysis.asymptotics
+
+/-!
+# Sub/sup-additive functions on boxes
+
+Let `s` be a set in `ι → ℝ`. A subbox of `s` is a product of closed intervals which is included
+by `s`.
+
+A function `f : (ι → ℝ) → (ι → ℝ) → M` is called
+`box_subadditive_on`/`box_additive_on`/`box_supadditive_on` a set `s` if for any `l u : ι → ℝ`, `l ≤
+u`, `Icc l u ⊆ s` and any hyperplane `x i = c`, `l i ≤ c ≤ u i`, the sum of the values of `f` on the
+two subboxes `[l, u] ∩ {x | x i ≤ c}` and `[l, u] ∩ {x | c ≤ x i}` is greater than or
+equal/equal/less than or equal to its value on `[l, u]`.
+
+The main result of this file is theorem `box_subadditive_on.eq_zero_of_forall_is_o_prod`. It says
+that `f l u = 0` provided that `f` is `box_subadditive_on` the interval `[l, u]`, `l ≤ u`, and
+for any `p ∈ [l, u]` we have `f l' u' = o(volume [l', u'])` as both `l'` tends to `p` along
+`[l, p]`, `u'` tends to `p` along `[p, u]`, and the subbox `[l', u']` is homothetic to `[l, u]`.
+-/
+
+variables {ι α β M : Type*}
+
+open set (univ pi Icc Ioi) function finset (hiding univ pi) filter
+open_locale big_operators topological_space nnreal filter
+
+/-!
+### Definitions and basic properties
+
+In this section we define `box_subadditive_on`, `box_additive_on`, and `box_supadditive_on`, and
+prove some basic properties.
+-/
+
+/-- A function `f : (ι → α) → (ι → α) → M` is called `box_subadditive_on` a set `s : set (ι → α)`
+if for any rectangular box `[l, u] ⊆ s` and a hyperplane `x i = c`, `l i ≤ c ≤ u i`,
+we have `f l u ≤ f' (I ∩ {x | x i ≤ c}) + f' (I ∩ {x | c ≤ x i})`, where `I = [l, u]`, and
```

Comment:
```suggestion
we have `f' I ≤ f' (I ∩ {x | x i ≤ c}) + f' (I ∩ {x | c ≤ x i})`, where `I = [l, u]`, and
```
also, the notation `[l, u]` is not completely standard to me in this higher-dimensional setting, so you could probably mention it in a notation section in the file-level docstring.